### PR TITLE
refactor natgrads to be more efficient

### PR DESCRIPTION
**Pull request by st--**
Previously, GPflow's NaturalGradient optimizer would call the loss_function once for each (q_mu, q_sqrt) set in the var_list. This is a light refactor that separates out applying the natural gradient step from computing the gradients (`_natgrad_apply_gradients`), and changes `_natgrad_steps` to only evaluate the loss function once, computing the gradients for all (q_mu, q_sqrt) tuples passed in the var_list.

Other changes:
- The no-longer-used `_natgrad_step` method got removed.
- NaturalGradient now takes a `xi_transform` argument that is used for all parameter sets without explicitly specified xi transform (i.e. tuples rather than triplets).
- XiTransform has been changed to have staticmethods.

None of this should affect any downstream code; this PR is backwards-compatible.

### CODE REVIEW

File: gpflow/optimizers/natgrad.py [hunk 1]
```diff
@@ -70,18 +69,22 @@ def naturals_to_xi(self, nat1, nat2):
         :param nat2: the θ₂ parameter
         :return: tuple `xi1`, `xi2`
         """
-        pass  # pragma: no cover
 
 
 class NaturalGradient(tf.optimizers.Optimizer):
-    def __init__(self, gamma: Scalar, name=None):
+    def __init__(self, gamma: Scalar, xi_transform: XiTransform = XiNat(), name=None):
```
Comment (awav) [inline gpflow/optimizers/natgrad.py:75]: Custom transformations go next to parameters and the can be different for a set of parameters. You cannot pass it in `init` method.

File: gpflow/optimizers/natgrad.py [hunk 2]
```diff
@@ -132,21 +178,29 @@ def _natgrad_step(
             year={2018}
 
         In addition, for convenience with the rest of GPflow, this code computes ∂L/∂η using
-        the chain rule:
+        the chain rule (the following assumes a numerator layout where the gradient is a row
+        vector; note that TensorFlow actually returns a column vector), where L is the loss:
 
-        ∂L/∂η = (∂L / ∂[q_μ, q_sqrt])(∂[q_μ, q_sqrt] / ∂η)
+        ∂L/∂η = (∂L / ∂[q_mu, q_sqrt])(∂[q_mu, q_sqrt] / ∂η)
```
Comment (awav) [inline gpflow/optimizers/natgrad.py:184]: Why is that changed?

File: gpflow/optimizers/natgrad.py (hunk 1 above)
Comment (condnsdmatters) [reply gpflow/optimizers/natgrad.py:115]: this is a default, it seems

Comment (condnsdmatters) [reply gpflow/optimizers/natgrad.py:115]: see line 101

Comment (st--) [reply gpflow/optimizers/natgrad.py:115]: This is simply changing the default when for a parameter you haven't passed in a transformation.

File: gpflow/optimizers/natgrad.py (hunk 2 above)
Comment (st--) [reply gpflow/optimizers/natgrad.py:219]: to be consistent with the code

Comment (condnsdmatters) [reply gpflow/optimizers/natgrad.py:219]: i think given this is the name of the variables we are using here, its kinda helpful to have it as q_mu. in general agree being against superfluous changes to docs for formatting but yeah

File: gpflow/optimizers/natgrad.py (hunk 1 above)
Comment (condnsdmatters) [reply gpflow/optimizers/natgrad.py:115]: Warning, this line will instantiate a static variable `XiNat` which will be shared by *all* instances of `NaturalGradient`.

I suggest changing to a class and instantiating.

```suggestion
    def __init__(self, gamma: Scalar, default_xi_transform: XiTransform = XiNat, name=None):
```
I know that currently the XiTransforms are stateless, but the abstractmethod APIs are not (they take self/arent classmethods)! So either we should say: you can define your own XiTransform which can take state (in which case, we should change this API), or you can not define your own stateful XiTransform  (and that should be reflected in API).

I suggest former.

File: gpflow/optimizers/natgrad.py (hunk 2 above)
Comment (st--) [reply gpflow/optimizers/natgrad.py:219]: you are against cleaning up the docs ? O.o

File: gpflow/optimizers/natgrad.py (hunk 1 above)
Comment (st--) [reply gpflow/optimizers/natgrad.py:115]: the XiNat transform does not have a state so it makes no difference if its instance is shared.

Comment (st--) [reply gpflow/optimizers/natgrad.py:115]: We can change the XiTransforms to be classmethods

**Comment by st--** [comment]
@condnsdmatters I've changed the XiTransform methods to all be static.

> _natgrad_step isn't used anywhere now?

no, the public API is just `minimize()`.

File: gpflow/optimizers/natgrad.py (hunk 1 above)
Comment (condnsdmatters) [reply gpflow/optimizers/natgrad.py:124]:
```suggestion
    def __init__(self, gamma: Scalar, xi_transform: XiTransform = XiNat, name=None):
```

File: gpflow/optimizers/natgrad.py (hunk 2 above)
Comment (awav) [reply gpflow/optimizers/natgrad.py:236]: I'm not against cleaning the docs. The cleaning could be done in different direction. I like q_μ more, and it doesn't contradicts the code.

File: gpflow/optimizers/natgrad.py [hunk 3]
```diff
@@ -39,114 +47,212 @@ class XiTransform(metaclass=abc.ABCMeta):
...
         Applies the transform so that `nat1`, `nat2` is mapped to `xi1`, `xi2`
 
         :param nat1: the θ₁ parameter
         :param nat2: the θ₂ parameter
         :return: tuple `xi1`, `xi2`
         """
-        pass  # pragma: no cover
+
+
+class XiNat(XiTransform):
```
Comment (awav) [inline gpflow/optimizers/natgrad.py:84]: I'm not sure. What happened here? :)

File: gpflow/optimizers/natgrad.py (hunk 1 above)
Comment (awav) [reply gpflow/optimizers/natgrad.py:144]: I'm sorry, but I still don't understand what transform does in `init`. It is "per parameter set" attribute. The default is `XiNat`, and if somebody wants to change it, he/she would need to specify it explicitly via parameters settings.

Comment (stefanosele) [reply gpflow/optimizers/natgrad.py:144]: What Artem want's to point out is that you may pass in let's say variational distributions of 3 layers (at least you could do that in the old code I think), and for some reason you want different transformation to be used for some of the parameters. You don't seem to be able to do that now. In general, the transformation should be tied to the parameter and not the optimizer

File: gpflow/optimizers/natgrad.py [hunk 4]
```diff
@@ -39,114 +47,223 @@ class XiTransform(metaclass=abc.ABCMeta):
...
         self.gamma = gamma
+        self.xi_transform = xi_transform
 
     def minimize(
-        self,
-        loss_fn: Callable,
-        var_list: List[
-            Union[Tuple[Parameter, Parameter], Tuple[Parameter, Parameter, XiTransform]]
-        ],
+        self, loss_fn: LossClosure, var_list: Sequence[NatGradParameters],
```
Comment (joelberkeley-pio) [inline gpflow/optimizers/natgrad.py:156]: i'm going to chime in about one thing cos I think it's important. This method signature isn't compatible with that of `tf.optimizer.Optimizer`. If that doesn't affect this PR, we can come back to it another time

Comment (st--) [reply gpflow/optimizers/natgrad.py:156]: Yes, very true! When reworking the code for this PR that is something that I also thought of. That is why I tried to make it much clearer in the docstrings - let me know if you have any further suggestions for clarifying that! I'm not sure what a good solution for this could be, I'd appreciate suggestions. It's beyond the scope of this PR though - I haven't actually changed the interface of this method, it behaves exactly as before.

File: gpflow/optimizers/natgrad.py (hunk 3 above)
Comment (st--) [reply gpflow/optimizers/natgrad.py:84]: I moved it to the top of the file so it could be used within the NaturalGradient class, and I changed the methods to be staticmethods as discussed with @condnsdmatters.

File: gpflow/optimizers/natgrad.py (hunk 2 above)
Comment (st--) [reply gpflow/optimizers/natgrad.py:243]: q_μ is consistent with neither paper notation (where it is just μ) nor the code (where it is q_mu), it took me extra mental effort to parse this as referring to q_mu...

File: gpflow/optimizers/natgrad.py (hunk 1 above)
Comment (st--) [reply gpflow/optimizers/natgrad.py:144]: You can still do that, just as before. I've not changed the interface (just minorly added to it, as I needed that for downstream code).

File: gpflow/optimizers/natgrad.py (hunk 4 above)
Comment (joelberkeley-pio) [reply gpflow/optimizers/natgrad.py:156]: perhaps it doesn't need to be a `tf.optimizers.Optimizer` and could just be its own class

to make it an actual `tf.optimizers.Optimizer`, I don't know how we'd do that. It may very well be possible but I couldn't say just yet

File: gpflow/optimizers/natgrad.py (hunk 1 above)
Comment (awav) [reply gpflow/optimizers/natgrad.py:144]: @st--, I think what I wanted to know is "why are you changing it"?

File: gpflow/optimizers/natgrad.py (hunk 2 above)
Comment (awav) [reply gpflow/optimizers/natgrad.py:243]: As you can see I had an opposite view. I would argue that it is better to keep consistency within one thing, rather trying to find consistency among everything. In this case, it was a documentation.

Comment (st--) [reply gpflow/optimizers/natgrad.py:243]: Sorry, did I not change it consistently - did I miss any occurrences? what is it that you think it's not consistent with, now?
